pkg/parquetquery/predicates.go

```diff
@@ -129,7 +140,7 @@ func (p *RegexInPredicate) keep(v *pq.Value) bool {
 
 	matched := false
 	for _, r := range p.regs {
-		if r.MatchString(s) {
+		if r.MatchString(s) == p.shouldMatch {
```

Comment:
I've been wondering if there is a performance benefit in using `r.Match()` instead.  Mostly because it would allow us to use `v.ByteArray()` instead of `v.String()`.   What do you think about creating a `BenchmarkRegexPredicate` and trying it out?

Reply:
With r.Match()
```
goos: darwin
goarch: arm64
pkg: github.com/grafana/tempo/pkg/parquetquery
BenchmarkRegexInPredicate-8   	   19226	     62875 ns/op
BenchmarkRegexInPredicate-8   	   18411	     64000 ns/op
BenchmarkRegexInPredicate-8   	   18625	     63660 ns/op
BenchmarkRegexInPredicate-8   	   19041	     66662 ns/op
BenchmarkRegexInPredicate-8   	   18672	     67335 ns/op
PASS
ok  	github.com/grafana/tempo/pkg/parquetquery	11.338s
```

with r.MatchString()
```
goos: darwin
goarch: arm64
pkg: github.com/grafana/tempo/pkg/parquetquery
BenchmarkRegexInPredicate-8   	   14379	     82846 ns/op
BenchmarkRegexInPredicate-8   	   14172	     82937 ns/op
BenchmarkRegexInPredicate-8   	   14335	     82151 ns/op
BenchmarkRegexInPredicate-8   	   14313	     85904 ns/op
BenchmarkRegexInPredicate-8   	   13190	     83861 ns/op
PASS
ok  	github.com/grafana/tempo/pkg/parquetquery	12.070s
```

Reply:
this is after turning off memoization

Reply:
* updated above with benchmem

Reply:
Ah my bad I forgot that the reason we use MatchString is because the memoization map key.  Shame because those benchmarks are nice.  Thanks for investigating.

Reply:
@mdisibio check this out.

Reply:
Thanks, I was just reading about the `map[string([]byte)]` lookup optimization as well.  My concern is that the byte slices from `v.ByteArray()` can't be held onto long-term because they point to the underlying page buffers and are reused ( unless you call `v.Clone()`).  So we need to dig into the go internals a bit and see if the string really does persist in the map or is just hashed.  Even if tests are passing, this might be tricky to catch.

Reply:
agree on that, I am also not much aware about it. I will look into `if the string really does persist in the map or is just hashed`

* also `v.String()` internally uses
```
	case ByteArray, FixedLenByteArray:
		return string(v.byteArray())
```

* and another impl, if we instead use `v.String()` for memoization and `v.ByteArray()` for matching. it still ~2x better than current implementation but ya the mem allocation will still be same as current one.
```
goos: darwin
goarch: arm64
pkg: github.com/grafana/tempo/pkg/parquetquery
BenchmarkRegexInPredicate-8   	   29047	     41418 ns/op	   48005 B/op	    1000 allocs/op
BenchmarkRegexInPredicate-8   	   28704	     41063 ns/op	   48005 B/op	    1000 allocs/op
BenchmarkRegexInPredicate-8   	   28605	     42268 ns/op	   48005 B/op	    1000 allocs/op
BenchmarkRegexInPredicate-8   	   29046	     41118 ns/op	   48005 B/op	    1000 allocs/op
BenchmarkRegexInPredicate-8   	   28389	     41852 ns/op	   48005 B/op	    1000 allocs/op
PASS
ok  	github.com/grafana/tempo/pkg/parquetquery	10.042s
```